operator/api/vectorized/v1alpha1/cluster_types.go

```diff
@@ -192,6 +198,45 @@ type ClusterSpec struct {
 	NodePools []NodePoolSpec `json:"nodePools,omitempty"`
 }
 
+// ClusterConfiguration holds values (or references to values) that should be used
+// to configure the cluster. Where the cluster schema defines a non-string type for a
+// given key, the corresponding values here should be string-encoded (according to yaml
+// rules)
+// TODO: this type should be lifted to a shared module rather than duplicated in the `redpanda` CRD definition.
+type ClusterConfiguration struct {
+	Values map[string]ClusterConfigValue `json:"values,omitempty"`
+}
+
+type YamlRepresentation string
+
+// ClusterConfigValue represents a value of arbitrary type T. Walues are string-encoded according to
+// YAML rules in order to preserve numerical fidelity.
+// Because these values must be embedded in a `.bootstrap.yaml` file - during the processing of
+// which, the AdminAPI's schema is unavailable - we endeavour to use yaml-compatible representations
+// throughout. The octet sequence of a Representation will be inserted into a bootstrap template
+// verbatim.
+// TODO: this type should be lifted to a shared module rather than duplicated in the `redpanda` CRD definition.
+type ClusterConfigValue struct {
+	// If the value is directly known, its yaml representation can be embedded here.
+	Representation *YamlRepresentation `json:"repr,omitempty"`
+	// If the value is supplied by an k8s object reference, coordinates are embedded here.
+	// For target values, the string value fetched from the source will be treated as
+	// a value encoded according to YAML rules; the string can then be embedded verbatim into
+	// the bootstrap file.
+	ConfigMapKeyRef *corev1.ConfigMapKeySelector `json:"configMapKeyRef,omitempty"`
+	// Should the value be contained in a k8s secret rather than configmap, we can refer
+	// to it here.
+	SecretKeyRef *corev1.SecretKeySelector `json:"secretKeyRef,omitempty"`
+	// Note: during the construction of a bootstrap template file, references to config objects above will be
+	// rewritten to an EnvVarRef instead, and that variable added to the container environment via the usual k8s
+	// envFrom mechanism. A preferred environment variable name may be specified here.
+	EnvVarRef *string `json:"envVar,omitempty"`
```

Comment:
I don't think we want to directly expose this, do we? We can rewrite the `{ConfigMap,Secret}KeyRef` as need be ourselves.

Reply:
This is down to me reusing the type, but I'll probably rejig that since the variations show up in different places.

Reply:
Fixed up.